doc/iplock.1
.TH IPLOCK 1 "August 2022" "iplock 1.x" "User Commands"
.SH NAME
iplock \- Block and Unblock IP addresses in iptables
.SH SYNOPSIS
.B iplock
\fICOMMAND\fR [\fIOPTIONS\fR] <ip1> <ip2> ... <ipN>
.SH DESCRIPTION
The \fBiplock(1)\fR tool is used to block and unblock IP addresses directly
whithin \fBiptables(8)\fR. The tool is setup so it is not necessary to be
root to add and remove IP addresses from the firewall. That allows the
\fBipwall(8)\fR service to add and remove IP addresses on all the computers
in a Snap! C++ cluster with one call.

The current implementation uses \fBipset(8)\fR to add and remove IPs from
a set. By default, the set is named \fBunwanted\fR. The administrator can
edit the `/etc/iplock/iplock.d/50-iplock.conf' file to change that name
or add multiple names by changing the value of the \fBallowed_sets\fR
variable.

.SH "SECURITY"
The \fBiplock(1)\fR tool only allows the update of very specific sets as
defined by the administrator. The names of the sets are defined in the
`/etc/iplock/iplock.conf' file and the `/etc/iplock/iplock.d/??-iplock.conf'
files as defined by other packages and the administrator. This is how the
tool can allow anyone to add and remove IP addresses from the set. It
does not allow the user from modifying the entire firewall.

The \-\-unblock command only removes the IP address from that very set.
It does not mean the IP will be allowed access since it may be blocked
by some other means (a different `ipset', a `recent' list, etc.)

.SH "DEBUGGING"
To help debug the tool, we offer the \fB\-\-verbose\fR command. This
displays the basic steps taken by \fBiplock(8)\fR before said step is
taken.

When adding (\fB\-\-block\fR) or removing (\fB\-\-unblock\fR) IPs to
a set, the tool creates a memory file which includes all the `add' or
`del' commands. To see that file, use the `\fB\-\-debug\fR' command
line option.

.SH "COMMANDS"
A set of options are viewed as commands specific to the \fBiplock(8)\fR
tool. These are listed here.

.TP
\fB\-b\fR, \fB\-\-block\fR
Block the list of specified IP addresses. The IP addresses can be listed
on the command line and in a file using the \fB\-\-ips\fR command line
option to define the filename.

The set being changed is the default one defined in the `iplock.conf'
configuration file. It can be changed on the command line using the
\fB\-\-set\fR command line option. Note that the name specified with
\fB\-\-set\fR must appear as one of the allowed sets (see the
`allowed_sets' parameter).

If an IP is already in the specified ipset, it is silently ignored.

If the IP address is one that appears in the `allowlist', or it matches
a network address part of the `allowlist', then that IP is not added to
the set. This is particularly useful to avoid having computers in a
cluster blocked by the firewall. It can also be used to allow third party
security checks by not blocking those third party servers.

.TP
\fB\-n\fR, \fB\-\-count\fR
Retrieve the current counters from a given \fBiptables(8)\fR chain. This
flag works with the \fB\-\-reset\fR and \fB\-\-total\fR command line options.

.TP
\fB\-f\fR, \fB\-\-flush\fR
Flush the set. This is similar to calling \fBiplock(8)\fR with
\fB\-\-unblock\fR for each one of the IP addresses currently being blocked,
just much faster and easier to use.

.TP
\fB\-l\fR, \fB\-\-list\fR
List the IP addresses currently blocked. This is the same as the
`ipset list ...' command for the specified set (see \fB\-\-set\fR).

Only the sets that the adminstrator allows can be listed in this way.
Other sets still require you to have enough permissions to do so and you
have to use the `ipset list ...' command directly.

.TP
\fB\-L\fR, \fB\-\-list\-allowed\-sets\fR
List the name of the sets that the administrator authorized \fBiplock(8)\fR
to access. When the \fB\-\-verbose\fR option is used, add `(*)' on the
line with the default set. By default, the project would output the following:

    $ iplock --list-allowed-sets
    unwanted (*)

.TP
\fB\-u\fR, \fB\-\-unblock\fR
Unblock a list of IP address as specified on the command line and in a file
as specified by the \fB\-\-ips\fR command line option.

The set the IPs are removed from can be changed using the \fB\-\-set\fR
command line option. By default, it is the "unwanted" set.

Nothing happens if the IP addresses are not currently present in the
specified set.

The `allowlist' is ignored when using this option. This is because you may
have added new IP addresses in the `allowlist' that were already added to
the set. It is important that \fBiplock(8)\fR allows the administrator to
still remove those IP addresses.


.SH "COMMAND LINE OPTIONS"
The following are all the other commands and options supported by
\fBiplock(8)\fR.

.TP
\fB\-\-build\-date\fR
Display the date and time when the tool was last built.

.TP
\fB\-\-command\-help\fR
List the commands understood by `iplock'.

.TP
\fB\-\-compiler\-version\fR
Output the version information of the C++ compiler used to compile the
`advgetopt' library. In most cases, this is the same as the compiler used
to compile the `iplock' tool.

.TP
\fB\-\-configuration\-filenames\fR
List the filename of each configuration file that the `iplock' will read
to get initialized.

\fBWARNING:\fR This does not include the setup files which get loaded
later in the process.

.TP
\fB\-\-console\fR
Print the logs out to the console. This overrides the logger configuration
files. By default, the iplock tool writes to the console if it is a tty.

.TP
\fB\-C\fR, \fB\-\-copyright\fR
Print out the copyright notice of the `iplock' tool.

.TP
\fB\-\-debug\fR
Change the logger severity to the `debug' level. This command line option
changes the level of all the appenders configured for `iplock'.

This is useful to fully debug the \fB\-\-block\fR and \fB\-\-unblock\fR
command line options. When in this mode, the memory file used to add
or remove the IPs from the set is printed to the console.

.TP
\fB\-\-environment\-variable\-name\fR
Print the name of the variable to the console. This variable can be used
to enter options instead of writing them on the command line or the
configuration file. Commands are not allowed in the environment variable.

.TP
\fB\-\-except\-stack\-collect\fR \fIno|yes|simple|complete\fR
Defines what exceptions are expected to collect from the stack.
Nearly all our exceptions are expected to stop a process. This command line
option gives you the ability to define how much information is to be collected
from the stack trace when that exception occurred. By default, it is set to
`yes' (which is the synomym of `simple'). With the logger, that exception
stack information can automatically be logged to your log file. Very useful
to debug issues in your software running on a remote server.

.TP
\fB\-\-force\-severity\fR \fIlevel\fR
Change the logger severity to this specific level. This new level is
applied to all the defined appenders. The level is changed whether it
is higher or lower than the current appender's level.

.TP
\fB\-\-has\-sanitizer\fR
Print whether this version was compiled with the C++ compiler sanitizer.

.TP
\fB\-h\fR, \fB\-\-help\fR
Print a brief document about the tool usage, then exit.

.TP
\fB\-\-ips\fR \fI<filename>\fR
Define the name of a file with IP addresses. This can be used along the
\fB\-\-block\fR and \fB\-\-unblock\fR commands.

The file format is simply IP addresses separated by commas, spaces, or
newlines. The `;' and `#' characters introduce a comment up to the next
newline character.

In most cases, these files are generated and only include one IP address
per line. It supports IPv4 and IPv6 addresses. For IPv6, it is safer to
write them between square brackets to make sure they are recognized as
IPv6 IPs.

.TP
\fB\-L\fR, \fB\-\-license\fR
Print out the license of `iplock' and exit.

.TP
\fB--list-appenders\fR
List the available appenders as used by the logger.

.TP
\fB\-\-list\-severities\fR
List the available severities as used by the logger.

.TP
\fB\-\-log\-component\fR \fIname\fR...
Define one or more component name to filter the logs. Only logs with that
component are output, others are ignored. You can use the `!' (exclamation
mark) character in front of a name to exclude logs with that component.

.TP
\fB\-\-log\-config\fR \fIfilename\fR
Only load this very configuration file to setup the logger.

.TP
\fB\-\-log\-config\-path\fR \fIpath\fR
Define a path to search for log configuration files instead of using the
defaults.

.TP
\fB\-\-log\-file\fR \fIfilename\fR
Save the logs in this specific file instead of what is defined in the
configuration files.

.TP
\fB\-\-logger\-configuration\-filenames\fR
List the files that the logger reads to be setup. If no files are found,
then messages appear in the console by default.

.TP
\fB\-\-logger\-hide\-banner\fR
Request that the logger does not print the banner in the console. This is
particularly useful for command line tools (opposed to a service). The
`iplock' tool automatically turns this option on.

.TP
\fB\-\-logger\-plugin\-paths\fR \fIpath1:path2:...\fR
This option allows you to change the path(s) to the `snaplogger' plugins. You
can enter multiple paths by separating them with a colon (:).

.TP
\fB\-\-logger\-show\-banner\fR
Request that the logger prints its banner when it starts. This is useful
if you run a tool from a service and want to make sure the administrator
knows which version of the tool is being used.

.TP
\fB\-\-logger\-version\fR
Print out the version of the Snap! Logger and exit.

.TP
\fB\-\-log\-severity\fR \fIlevel\fR
Change the severity to this level unless the appender already accepts logs
for a lower level. So for example, an appender setup to output logs with a
severity level of `error' or higher is affected by the `warning' level.
An appender setup to output logs with a severity level of `info' does not
changed when setting this command line option to `warning'.

.TP
\fB\-\-no\-log\fR
Turn off the logger so nothing gets printed out. This is somewhat similar
to a quiet or silent option that many Unix tools offer.

.TP
\fB\-\-option\-help\fR
Print the list of options supported by `iplock'.

.TP
\fB\-\-path\-to\-option\-definitions\fR
Option definitions can be defined in a .ini file. If it exists, this is the
path where it can be found.

.TP
\fB\-\-print\-option\fR \fIname\fR
This option is useful to debug your command line, environment variable, and
configuration files by printing the value of the named option. This tells
you what value the tool or service sees in the end.

.TP
\fB\-\-quiet\fR
Turn off all messages, including errors, from the \fBiptables(8)\fR and
\fBipset(8)\fR tools. Use this when running such commands from other
software to avoid generally unnecessary output.

.TP
\fB\-r\fR, \fB\-\-reset\fR
Use this flag along the \fB\-\-count\fR command line option. It requests
the command to also reset the counter. This is useful if you are gathering
the counters to save them in a database to use as statistics.

.TP
\fB\-s\fR, \fB\-\-set\fR \fI<name>\fR
Change the name of the set on which to work. The internal default is
"unwanted". Only sets that the administrator allowed can be specified
here. Other names, even if the set exists, generate a fatal error.

.TP
\fB\-\-show\-option\-sources\fR
The `advgetopt' library has the ability to trace where each value is
read from: which file, the command line, the environment variable.
This option can be used to determine where a value is defined, which once
in a while is particularly useful.

.TP
\fB\-\-syslog\fR [\fIidentity\fR]
Send the logs to the system `syslog'. If specified, the `identity' is used
along each message.

.TP
\fB\-\-trace\fR
Change the logger severity to the TRACE level. All appenders accept all the
logs that they receive.

.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print information about what the tool is doing. In particular, it shows
the commands that are run to help debug issue if such arise.

.TP
\fB\-V\fR, \fB\-\-version\fR
Print the `iplock' version number, then exit.

.SH AUTHOR
Written by Alexis Wilke <alexis@m2osw.com>.
.SH "REPORTING BUGS"
Report bugs to <https://github.com/m2osw/iplock/issues>.
.br
iplock home page: <https://snapwebsites.org/project/iplock>.
.SH COPYRIGHT
Copyright \(co 2022-2026  Made to Order Software Corp.  All Rights Reserved
.br
License: GPLv3
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ipload (5),
.BR ipload (8),
.BR iprecent (1),
.BR ipwall (8),
.BR knock-knock (1),
.BR showfw (1)


debian/changelog
iplock (3.0.57.1~noble) noble; urgency=high

  * Fixed ipwall service description. The boot part is ipload.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 26 Jul 2025 11:11:35 -0700

iplock (3.0.57.0~noble) noble; urgency=high

  * ipload now prints a message on success which will make the logs
    easier to read (otherwise it is not 100% clear that it succeeded).
  * The default ipload firewall was created in a non-existent folder
    (/run/users/0,) this version will instead use /var/lib/iplock.
  * Do not save the showfw output in /run/user/0/...; use the /tmp/iplock
    folder instead (create if necessary using umask for security).
  * Renamed get_port_defined() as is_port_defined().
  * Fixed the service so it does not depend on the default dependencies.
  * Changed the BLOCK and UNBLOCK messages with IPWALL_... prefix.
  * Enhanced the IPWALL_CURRENT_STATUS and firewall UP concepts.
  * Detect all interfaces, especially those that are up if not expected.
  * Properly handle the STATUS message with a dispatch CALLBACK.
  * Fixed a [verify] section start point.
  * Fixed some documentation, old references, etc.
  * Updated the debian/compat to use the dependency instead.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 21 Jul 2024 15:04:06 -0700

iplock (3.0.56.0~bionic) bionic; urgency=high

  * Changed the restart of rsyslogd using systemd.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 06 Jan 2024 14:21:44 -0800

iplock (3.0.55.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 21 Dec 2023 21:48:35 -0800

iplock (3.0.55.0~bionic) bionic; urgency=high

  * Made our config dpkg-reconfigure compatible.

 -- Alexis Wilke <alexis@m2osw.com>  Thu, 07 Dec 2023 20:13:50 -0800

iplock (3.0.54.0~bionic) bionic; urgency=high

  * Fix config to use `db_settitle`.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 06 Dec 2023 21:47:31 -0800

iplock (3.0.53.0~bionic) bionic; urgency=high

  * Adding example to admin_interfaces parameter.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 05 Dec 2023 20:46:46 -0800

iplock (3.0.52.0~bionic) bionic; urgency=high

  * Actually output admin_interfaces to 00-variables.conf.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 04 Dec 2023 21:15:36 -0800

iplock (3.0.51.0~bionic) bionic; urgency=high

  * Added admin_interfaces to the iplock.config file.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 04 Dec 2023 17:59:43 -0800

iplock (3.0.50.0~bionic) bionic; urgency=high

  * Moved the admin_interfaces & admin_ips variables to the variables.conf file.
  * Make use of the admin_interfaces in two more locations.
  * Added admin_interfaces to the config template.

 -- Alexis Wilke <alexis@m2osw.com>  Mon, 04 Dec 2023 17:41:59 -0800

iplock (3.0.49.0~bionic) bionic; urgency=high

  * Fixed a missing quote (") in an expression in the ipload conf.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 03 Dec 2023 21:33:33 -0800

iplock (3.0.48.0~bionic) bionic; urgency=high

  * Allow for the admin_ips to be "0.0.0.0/0".

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 18 Nov 2023 17:48:06 -0800

iplock (3.0.47.0~bionic) bionic; urgency=high

  * Avoid smashing the "public_interfaces" value if already defined.
  * Extended the `remove_user_chain` script to also reset the IPv6 tables.
  * Use the `POJECTNAME` variable in one more location in postinst.
  * Small clean ups.

 -- Alexis Wilke <alexis@m2osw.com>  Sat, 18 Nov 2023 10:43:50 -0800

iplock (3.0.46.3~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 17 Nov 2023 18:28:37 -0800

iplock (3.0.46.2~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 15 Nov 2023 18:28:52 -0800

iplock (3.0.46.1~bionic) bionic; urgency=high

  * Bumped build version to rebuild on Launchpad.

 -- Alexis Wilke <alexis@m2osw.com>  Wed, 15 Nov 2023 11:37:34 -0800

iplock (3.0.46.0~bionic) bionic; urgency=high

  * Fixed postinst script so it installs the ${admin_ips}.
  * Fixed path to the README.md file.

 -- Alexis Wilke <alexis@m2osw.com>  Tue, 14 Nov 2023 21:41:23 -0800

iplock (3.0.45.0~bionic) bionic; urgency=high

  * Added flag to turn syn-flood ON or OFF. It is OFF by default for now.
  * Added info about the possibility to use `-m limit` for synflood.
  * Use dh_installsystemd instead of directly using systemctl for ipload.
  * Fixed an #include <cstdint>.
  * Removed an std::move() which did not work anymore in newer g++ versions.
  * Install the log images.
  * Updated compat to version 15.
  * Fixed packages
  * Fixed ipwall so ipload gets installed & started as expected.
  * Tweaked the logrotate post script.
  * Make use of names.an instead of hard coded strings.
  * Fixed a link in README.md.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 20 Nov 2022 21:23:07 -0800

iplock (3.0.44.0~bionic) bionic; urgency=high

  * Put 127.0.0.0/8 as the default allowed list.

 -- Alexis Wilke <alexis@m2osw.com>  Fri, 18 Nov 2022 20:45:37 -0800

iplock (3.0.43.0~bionic) bionic; urgency=high

  * Removing the "broken_tcp" rules. They catch nothing at all.
  * Started looking in nftables. I think we'll need that within 2 or 3 years.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 13 Nov 2022 21:07:58 -0800

iplock (3.0.42.0~bionic) bionic; urgency=high

  * Fixed the synflood by doing the reap on its own without the --hitcount.

 -- Alexis Wilke <alexis@m2osw.com>  Sun, 13 Nov 2022 17:00:19 -0800

iplock (3.0.41.0~bionic) bionic; urgency=high

  * Added support for the extended assignment operators within the iplock.conf.
  * When there are multiple recent it's for a single rule.
  * Output the TCP state before the -m recent rules.
  * Reverted the synflood hitcount and duration to 100 / 60s.
  * Moved the synflood to the unwanted chain after SSH.
  * Fixed the comment (documentation) of the synflood rule.
  * Fixed the knock-knock TCP checks vs. Recent rules:

tools/iplock/iplock.cpp
// Copyright (c) 2014-2022  Made to Order Software Corp.  All Rights Reserved
//
// https://snapwebsites.org/project/iplock
// contact@m2osw.com
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.


/** \file
 * \brief iplock tool.
 *
 * This implementation offers a way to easily and safely add and remove
 * IP addresses one wants to block/unblock temporarily.
 *
 * The tool makes use of the iptables tool to add and remove rules
 * to one specific table which is expected to be included in your
 * INPUT rules (with a `-j \<table-name>`).
 */


// self
//
#include    "iplock.h"

#include    "batch.h"
#include    "block.h"
#include    "count.h"
#include    "flush.h"
#include    "unblock.h"


// iplock
//
#include    <iplock/version.h>


// boost
//
#include    <boost/preprocessor/stringize.hpp>


// C++
//
#include    <iostream>


// C
//
#include    <unistd.h>


// last include
//
#include    <snapdev/poison.h>



namespace tool
{


/** \brief Command line options.
 *
 * This table includes all the options supported by iplock on the
 * command line.
 */
advgetopt::option const g_iplock_options[] =
{
    // COMMANDS
    //
    advgetopt::define_option(
          advgetopt::Name("batch")
        , advgetopt::ShortName('a')
        , advgetopt::Flags(advgetopt::command_flags<
                      advgetopt::GETOPT_FLAG_GROUP_COMMANDS
                    , advgetopt::GETOPT_FLAG_REQUIRED>())
        , advgetopt::Help("Text file containing rules to add to the firewall.")
    ),
    advgetopt::define_option(
          advgetopt::Name("block")
        , advgetopt::ShortName('b')
        , advgetopt::Flags(advgetopt::standalone_command_flags<
                      advgetopt::GETOPT_FLAG_GROUP_COMMANDS>())
        , advgetopt::Help("Block the speficied IP address. If already blocked, do nothing.")
    ),
    advgetopt::define_option(
          advgetopt::Name("count")
        , advgetopt::ShortName('n')
        , advgetopt::Flags(advgetopt::standalone_command_flags<
                      advgetopt::GETOPT_FLAG_GROUP_COMMANDS>())
        , advgetopt::Help("Return the number of times each IP address was"
                " blocked since the last counter reset. You may use the"
                " --reset along this command to atomically reset the"
                " counters as you retrieve them.")
    ),
    advgetopt::define_option(
          advgetopt::Name("flush")
        , advgetopt::ShortName('f')
        , advgetopt::Flags(advgetopt::standalone_command_flags<
                      advgetopt::GETOPT_FLAG_GROUP_COMMANDS
                    , advgetopt::GETOPT_FLAG_SHOW_USAGE_ON_ERROR>())
        , advgetopt::Help("Flush all rules specified in chain.")
    ),
    advgetopt::define_option(
          advgetopt::Name("unblock")
        , advgetopt::ShortName('u')
        , advgetopt::Flags(advgetopt::option_flags<
                      advgetopt::GETOPT_FLAG_GROUP_COMMANDS
                    , advgetopt::GETOPT_FLAG_COMMAND_LINE>())
        , advgetopt::Help("Unblock the specified IP address. If not already blocked, do nothing.")
    ),

    // OPTIONS
    //
    advgetopt::define_option(
          advgetopt::Name("quiet")
        , advgetopt::ShortName('q')
        , advgetopt::Flags(advgetopt::option_flags<
                      advgetopt::GETOPT_FLAG_GROUP_OPTIONS
                    , advgetopt::GETOPT_FLAG_COMMAND_LINE
                    , advgetopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE>())
        , advgetopt::Help("Prevent iptables from printing messages in stdout or stderr.")
    ),
    advgetopt::define_option(
          advgetopt::Name("reset")
        , advgetopt::ShortName('r')
        , advgetopt::Flags(advgetopt::option_flags<
                      advgetopt::GETOPT_FLAG_GROUP_OPTIONS
                    , advgetopt::GETOPT_FLAG_COMMAND_LINE
                    , advgetopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE>())
        , advgetopt::Help("Use with the --count command to retrieve the counters and reset them atomically.")
    ),
    advgetopt::define_option(
          advgetopt::Name("scheme")
        , advgetopt::ShortName('s')
        , advgetopt::Flags(advgetopt::any_flags<
                      advgetopt::GETOPT_FLAG_GROUP_OPTIONS
                    , advgetopt::GETOPT_FLAG_COMMAND_LINE
                    , advgetopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE
                    , advgetopt::GETOPT_FLAG_REQUIRED>())
        , advgetopt::Help("Configuration file to define iptables commands. This is one name (no '/' or '.'). The default is \"http\".")
    ),
    advgetopt::define_option(
          advgetopt::Name("total")
        , advgetopt::ShortName('t')
        , advgetopt::Flags(advgetopt::option_flags<
                      advgetopt::GETOPT_FLAG_GROUP_OPTIONS
                    , advgetopt::GETOPT_FLAG_COMMAND_LINE
                    , advgetopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE>())
        , advgetopt::Help("Write the grand total only when --count is specified.")
    ),
    advgetopt::define_option(
          advgetopt::Name("verbose")
        , advgetopt::ShortName('v')
        , advgetopt::Flags(advgetopt::option_flags<
                      advgetopt::GETOPT_FLAG_GROUP_OPTIONS
                    , advgetopt::GETOPT_FLAG_COMMAND_LINE
                    , advgetopt::GETOPT_FLAG_ENVIRONMENT_VARIABLE>())
        , advgetopt::Help("Show commands being executed.")
    ),
    advgetopt::define_option(
          advgetopt::Name("--")
        , advgetopt::Flags(advgetopt::command_flags<
                      advgetopt::GETOPT_FLAG_GROUP_OPTIONS
                    , advgetopt::GETOPT_FLAG_DEFAULT_OPTION
                    , advgetopt::GETOPT_FLAG_MULTIPLE>())
    ),
    advgetopt::end_options()
};



advgetopt::group_description const g_group_descriptions[] =
{
    advgetopt::define_group(
          advgetopt::GroupNumber(advgetopt::GETOPT_FLAG_GROUP_COMMANDS)
        , advgetopt::GroupName("command")
        , advgetopt::GroupDescription("Commands:")
    ),
    advgetopt::define_group(
          advgetopt::GroupNumber(advgetopt::GETOPT_FLAG_GROUP_OPTIONS)
        , advgetopt::GroupName("option")
        , advgetopt::GroupDescription("Options:")
    ),
    advgetopt::end_groups()
};




// TODO: once we have stdc++20, remove all defaults
#pragma GCC diagnostic ignored "-Wpedantic"
advgetopt::options_environment const g_iplock_options_environment =
{
    .f_project_name = "iplock",
    .f_group_name = nullptr,
    .f_options = g_iplock_options,
    .f_options_files_directory = nullptr,
    .f_environment_variable_name = "IPLOCK_OPTIONS",
    .f_environment_variable_intro = nullptr,
    .f_section_variables_name = nullptr,
    .f_configuration_files = nullptr,
    .f_configuration_filename = nullptr,
    .f_configuration_directories = nullptr,
    .f_environment_flags = advgetopt::GETOPT_ENVIRONMENT_FLAG_SYSTEM_PARAMETERS
                         | advgetopt::GETOPT_ENVIRONMENT_FLAG_PROCESS_SYSTEM_PARAMETERS,
    .f_help_header = "Usage: %p [-<opt>] [ip]\n"
                     "where -<opt> is one or more of:",
    .f_help_footer = nullptr,
    .f_version = IPLOCK_VERSION_STRING,
    .f_license = "GNU GPL 3",
    .f_copyright = "Copyright (c) 2007-" BOOST_PP_STRINGIZE(UTC_BUILD_YEAR) " by Made to Order Software Corporation",
    .f_build_date = UTC_BUILD_DATE,
    .f_build_time = UTC_BUILD_TIME,
    .f_groups = g_group_descriptions
};






/** \brief Initialize the iplock object.
 *
 * This function parses the command line and  determines the command
 * that the end user selected (i.e. --block, --unblock, or --count.)
 *
 * If the user specified --help or --version, then this function
 * prints the help screen or version of iplock and exits the process
 * immediately.
 *
 * If no command was specified on the command line, then an error
 * is written to stderr and the process exits immediately.
 *
 * \param[in] argc  The number of arguments in argv.
 * \param[in] argv  The argument strings.
 */
iplock::iplock(int argc, char * argv[])
{
    advgetopt::getopt::pointer_t opts(std::make_shared<advgetopt::getopt>(g_iplock_options_environment, argc, argv));

    // define the command
    //
    // since the user may specify any number of commands, we use
    // the set_command() function to make sure that only one
    // gets set...
    //
    if(opts->is_defined("block"))
    {
        set_command(std::make_shared<block>(this, opts));
    }
    if(opts->is_defined("unblock"))
    {
        set_command(std::make_shared<unblock>(this, opts));
    }
    if(opts->is_defined("count"))
    {
        set_command(std::make_shared<count>(this, opts));
    }
    if(opts->is_defined("flush"))
    {
        set_command(std::make_shared<flush>(this, opts));
    }
    if(opts->is_defined("batch"))
    {
        set_command(std::make_shared<batch>(this, opts));
    }

    // no command specified?
    //
    if(f_command == nullptr)
    {
        std::cerr << "iplock:error: you must specify one of: --block, --unblock, --count, --flush, or --batch.\n";
        exit(1);
    }
}


/** \brief Save the command pointer in f_command.
 *
 * This function saves the specified \p c command pointer to the f_command
 * parameter.
 *
 * It is done that way so we can easily detect whether more than one
 * command was specified on the command line.
 *
 * \param[in] c  The pointer to the command line to save in iplock.
 */
void iplock::set_command(command::pointer_t c)
{
    if(f_command != nullptr)
    {
        std::cerr << "iplock:error: you can only specify one command at a time, one of: --block, --unblock, --count, --flush, or --batch." << std::endl;
        exit(1);
    }
    f_command = c;
}


/** \brief Before running a command, make sure we are root.
 *
 * This function gets called by the run_command() function.
 *
 * The function exits the process with an error if becoming root is not
 * possible. This can happen if (1) the process is run by systemd and
 * systemd prevents such, (2) the binary is not marked with the 's'
 * bit.
 */
void iplock::make_root()
{
    if(setuid(0) != 0)
    {
        perror("iplock:error: setuid(0)");
        exit(1);
    }
    if(setgid(0) != 0)
    {
        perror("iplock:error: setgid(0)");
        exit(1);
    }
}


/** \brief Run the selected command.
 *
 * The constructor parses the command line options and from that
 * deterimes which command the user selected. This function runs
 * that command by calling its run() function.
 *
 * This function first makes sure the user is running as root.
 * This may change in the future if some of the commands may
 * otherwise be run as a regular user.
 */
int iplock::run_command()
{
    // all iptables commands require the user to be root.
    //
    make_root();

    f_command->run();

    return f_command->exit_code();
}



} // namespace tool
// vim: ts=4 sw=4 et
